substrate-archive/src/wasm_tracing.rs

```diff
@@ -0,0 +1,329 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with substrate-archive.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The Subscriber implementation for Tracing
+
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::sync::{
+	atomic::{AtomicU64, Ordering},
+	Arc,
+};
+
+use chrono::{DateTime, Utc};
+use parking_lot::Mutex;
+use serde::{Deserialize, Serialize};
+use tracing::{
+	event::Event,
+	field::{Field, Visit},
+	span::{Attributes, Id, Record},
+	Level, Metadata, Subscriber,
+};
+use tracing_subscriber::CurrentSpan;
+
+use sp_tracing::{WASM_NAME_KEY, WASM_TARGET_KEY, WASM_TRACE_IDENTIFIER};
+use substrate_archive_common::{Result, TracingError};
+
+/// Generic BlockHash type that can be any length.
+type BlockHash = Vec<u8>;
+
+/// The Event a tracing subscriber collects before sending data to the TracingActor.
+#[derive(Debug)]
+pub struct EventMessage {
+	pub block_num: Option<u32>,
+	pub hash: Option<Vec<u8>>,
+	pub name: String,
+	pub target: String,
+	pub level: Level,
+	pub values: TraceData,
+	pub parent_id: Id,
+	pub time: DateTime<Utc>,
+	pub file: Option<String>,
+	pub line: Option<u32>,
+}
+
+/// The message a tracing subscriber collects before sending data to the TracingActor.
+#[derive(Debug, Clone)]
+pub struct SpanMessage {
+	pub id: Id,
+	pub parent_id: Option<Id>,
+	pub name: String,
+	pub target: String,
+	pub level: Level,
+	pub values: TraceData,
+	pub start_time: DateTime<Utc>,
+	pub overall_time: chrono::Duration,
+	pub file: Option<String>,
+	pub line: Option<u32>,
+}
+
+/// Finished Trace Data Format. Ready for insertion into a relational database.
+#[derive(Debug)]
+pub struct Traces {
+	block_num: u32,
+	hash: Vec<u8>,
+	pub spans: Vec<SpanMessage>,
+	pub events: Vec<EventMessage>,
+}
+
+impl Traces {
+	pub fn new(block_num: u32, hash: Vec<u8>, events: Vec<EventMessage>, spans: Vec<SpanMessage>) -> Self {
+		Traces { block_num, hash, spans, events }
+	}
+
+	/// Get the hash these spans come from.
+	pub fn hash(&self) -> Vec<u8> {
+		self.hash.clone()
+	}
+
+	/// Get the block number of the block these spans come from.
+	pub fn block_num(&self) -> u32 {
+		self.block_num
+	}
+}
+
+#[derive(Debug)]
+pub struct SpanEvents {
+	pub spans: Vec<SpanMessage>,
+	pub events: Vec<EventMessage>,
+}
+
+/// Collects traces and filters based on target.
+/// The Subscriber implementation is blocking. It uses Mutex primitives to coalesce traces before
+/// sending them to the appropriate actor.
+/// Therefore, one must be careful not to block the async executor when adding tracing spans
+/// using this subscriber implementation anywhere inside an async context in substrate-archive.
+pub struct TraceHandler {
+	/// the Block Number this trace set is from
+	block_num: u32,
+	/// Hash of the block for this trace set
+	hash: BlockHash,
+	span_events: Arc<Mutex<SpanEvents>>,
```

Comment:
This seems OK - but alternatively maybe put Spans and Events in separately behind their own Mutex?

Reply:
Yeah -- not sure if there would be much benefit since the runtime is single threaded anyway so we shouldn't hit this mutex iin different threads at the same time very much; it's cleaner to instantiate this way imo (this is mostly personal preference though. If it is actually better to use two mutexes I can do that)

Reply:
There are many interesting performance questions around this PR and I think we could run a good number of experiments that are interesting to the wider substrate team. At this time though I'm more interested in delivering the state-tracing feature and so I'm deliberately holding back the itch to benchmark stuff.

So yeah, I'm fine keeping a single mutex for now. :)

Reply:
I think at some point it would also be interesting to benchmark other parts of archive too, and optimize some of our defaults (block load/threads/wasm cache etc)